File: Airship/AirshipCore/Source/Internal/UARetailEventTemplate.m
/* Copyright Airship and Contributors */

#import "UARetailEventTemplate.h"
#import "UAirship.h"
#import "UAAnalytics.h"
#import "UACustomEvent+Internal.h"

#define kUARetailEventTemplate @"retail"
#define kUABrowsedProductEvent @"browsed"
#define kUAAddedToCartEvent @"added_to_cart"
#define kUAStarredProductEvent @"starred_product"
#define kUASharedProductEvent @"shared_product"
#define kUAPurchasedEvent @"purchased"
#define kUAWishlistEvent @"wishlist"
#define kUARetailEventTemplateLifetimeValue @"ltv"
#define kUARetailEventTemplateIdentifier @"id"
#define kUARetailEventTemplateCategory @"category"
#define kUARetailEventTemplateDescription @"description"
#define kUARetailEventTemplateBrand @"brand"
#define kUARetailEventTemplateNewItem @"new_item"
#define kUARetailEventTemplateSource @"source"
#define kUARetailEventTemplateMedium @"medium"
#define kUARetailEventTemplateWishlistName @"wishlist_name"
#define kUARetailEventTemplateWishlistID @"wishlist_id"

@interface UARetailEventTemplate()
@property (nonatomic, copy) NSString *eventName;
@property (nonatomic, copy) NSString *source;
@property (nonatomic, copy) NSString *medium;
@property (nonatomic, copy) NSString *wishlistName;
@property (nonatomic, copy) NSString *wishlistID;
@property (nonatomic, assign) BOOL newItemSet;
@end

@implementation UARetailEventTemplate

- (instancetype)initWithName:(NSString *)name
                   withValue:(NSDecimalNumber *)eventValue
                  withSource:(NSString *)source
                  withMedium:(NSString *)medium
            withWishlistName:(NSString *)wishlistName
              withWishlistID:(NSString *)wishlistID {
    self = [super init];
    if (self) {
        self.eventName = name;
        self.eventValue = eventValue;
        self.source = source;
        self.medium = medium;
        self.wishlistName = wishlistName;
        self.wishlistID = wishlistID;
    }

    return self;
}

+ (instancetype)browsedTemplate {
    return [self browsedTemplateWithValue:nil];
}

+ (instancetype)browsedTemplateWithValueFromString:(NSString *)eventValue {
    NSDecimalNumber *decimalValue = eventValue ? [NSDecimalNumber decimalNumberWithString:eventValue] : nil;
    return [self browsedTemplateWithValue:decimalValue];
}

+ (instancetype)browsedTemplateWithValue:(NSDecimalNumber *)eventValue {
    return [[self alloc] initWithName:kUABrowsedProductEvent
                            withValue:eventValue
                           withSource:nil
                           withMedium:nil
                     withWishlistName:nil
                       withWishlistID:nil];
}

+ (instancetype)addedToCartTemplate {
    return [self addedToCartTemplateWithValue:nil];
}

+ (instancetype)addedToCartTemplateWithValueFromString:(NSString *)eventValue {
    NSDecimalNumber *decimalValue = eventValue ? [NSDecimalNumber decimalNumberWithString:eventValue] : nil;
    return [self addedToCartTemplateWithValue:decimalValue];
}

+ (instancetype)addedToCartTemplateWithValue:(NSDecimalNumber *)eventValue {
    return [[self alloc] initWithName:kUAAddedToCartEvent
                            withValue:eventValue
                           withSource:nil
                           withMedium:nil
                     withWishlistName:nil
                       withWishlistID:nil];
}

+ (instancetype)starredProductTemplate {
    return [self starredProductTemplateWithValue:nil];
}

+ (instancetype)starredProductTemplateWithValueFromString:(NSString *)eventValue {
    NSDecimalNumber *decimalValue = eventValue ? [NSDecimalNumber decimalNumberWithString:eventValue] : nil;
    return [self starredProductTemplateWithValue:decimalValue];
}

+ (instancetype)starredProductTemplateWithValue:(NSDecimalNumber *)eventValue {
    return [[self alloc] initWithName:kUAStarredProductEvent
                            withValue:eventValue
                           withSource:nil
                           withMedium:nil
                     withWishlistName:nil
                       withWishlistID:nil];
}

+ (instancetype)purchasedTemplate {
    return [self purchasedTemplateWithValue:nil];
}

+ (instancetype)purchasedTemplateWithValueFromString:(NSString *)eventValue {
    NSDecimalNumber *decimalValue = eventValue ? [NSDecimalNumber decimalNumberWithString:eventValue] : nil;
    return [self purchasedTemplateWithValue:decimalValue];
}

+ (instancetype)purchasedTemplateWithValue:(NSDecimalNumber *)eventValue {
    return [[self alloc] initWithName:kUAPurchasedEvent
                            withValue:eventValue
                           withSource:nil
                           withMedium:nil
                     withWishlistName:nil
                       withWishlistID:nil];
}

+ (instancetype)sharedProductTemplate {
    return [self sharedProductTemplateWithValue:nil];
}

+ (instancetype)sharedProductTemplateWithValueFromString:(NSString *)eventValue {
    NSDecimalNumber *decimalValue = eventValue ? [NSDecimalNumber decimalNumberWithString:eventValue] : nil;
    return [self sharedProductTemplateWithValue:decimalValue];
}

+ (instancetype)sharedProductTemplateWithValue:(NSDecimalNumber *)eventValue {
    return [[self alloc] initWithName:kUASharedProductEvent
                            withValue:eventValue
                           withSource:nil
                           withMedium:nil
                     withWishlistName:nil
                       withWishlistID:nil];
}

+ (instancetype)sharedProductTemplateWithSource:(NSString *)source
                                  withMedium:(NSString *)medium {
    return [[self alloc] initWithName:kUASharedProductEvent
                            withValue:nil
                           withSource:source
                           withMedium:medium
                     withWishlistName:nil
                       withWishlistID:nil];
}

+ (instancetype)sharedProductTemplateWithValueFromString:(NSString *)eventValue
                                           withSource:(NSString *)source
                                           withMedium:(NSString *)medium {
    NSDecimalNumber *decimalValue = eventValue ? [NSDecimalNumber decimalNumberWithString:eventValue] : nil;
    return [[self alloc] initWithName:kUASharedProductEvent
                            withValue:decimalValue
                           withSource:source
                           withMedium:medium
                     withWishlistName:nil
                       withWishlistID:nil];
}

+ (instancetype)sharedProductTemplateWithValue:(NSDecimalNumber *)eventValue
                                 withSource:(NSString *)source
                                 withMedium:(NSString *)medium {
    return [[self alloc] initWithName:kUASharedProductEvent
                            withValue:eventValue
                           withSource:source
                           withMedium:medium
                     withWishlistName:nil
                       withWishlistID:nil];
}

+ (instancetype)wishlistTemplate {
    return [[self alloc] initWithName:kUAWishlistEvent
                            withValue:nil
                           withSource:nil
                           withMedium:nil
                     withWishlistName:nil
                       withWishlistID:nil];
}

+ (instancetype)wishlistTemplateWithName:(NSString *)name wishlistID:(NSString *)wishlistID {
    return [[self alloc] initWithName:kUAWishlistEvent
                            withValue:nil
                           withSource:nil
                           withMedium:nil
                     withWishlistName:name
                       withWishlistID:wishlistID];
}

- (void)setEventValue:(NSDecimalNumber *)eventValue {
    if (!eventValue) {
        _eventValue = nil;
    } else {
        if ([eventValue isKindOfClass:[NSDecimalNumber class]]) {
            _eventValue = eventValue;
        } else {
            _eventValue = [NSDecimalNumber decimalNumberWithDecimal:[eventValue decimalValue]];
        }
    }
}

- (void)setIsNewItem:(BOOL)isNewItem {
    self.newItemSet = YES;
    _isNewItem = isNewItem;
}

- (UACustomEvent *)createEvent {
    UACustomEvent *event = [UACustomEvent eventWithName:self.eventName];

    NSMutableDictionary *propertyDictionary = [NSMutableDictionary dictionary];
    
    if (self.eventValue) {
        [event setEventValue:self.eventValue];
    }

    if (self.eventValue && [self.eventName isEqualToString:kUAPurchasedEvent]) {
        [propertyDictionary setValue:@YES forKey:kUARetailEventTemplateLifetimeValue];
    } else {
       [propertyDictionary setValue:@NO forKey:kUARetailEventTemplateLifetimeValue];
    }

    if (self.transactionID) {
        [event setTransactionID:self.transactionID];
    }

    if (self.identifier) {
        [propertyDictionary setValue:self.identifier forKey:kUARetailEventTemplateIdentifier];
    }

    if (self.category) {
        [propertyDictionary setValue:self.category forKey:kUARetailEventTemplateCategory];
    }

    if (self.eventDescription) {
        [propertyDictionary setValue:self.eventDescription forKey:kUARetailEventTemplateDescription];
    }

    if (self.brand) {
        [propertyDictionary setValue:self.brand forKey:kUARetailEventTemplateBrand];
    }

    if (self.newItemSet) {
        [propertyDictionary setValue:@(self.isNewItem) forKey:kUARetailEventTemplateNewItem];
    }
    if (self.source) {
        [propertyDictionary setValue:self.source forKey:kUARetailEventTemplateSource];
    }

    if (self.medium) {
        [propertyDictionary setValue:self.medium forKey:kUARetailEventTemplateMedium];
    }
    
    if (self.wishlistID) {
        [propertyDictionary setValue:self.wishlistID forKey:kUARetailEventTemplateWishlistID];
    }
    
    if (self.wishlistName) {
        [propertyDictionary setValue:self.wishlistName forKey:kUARetailEventTemplateWishlistName];
    }

    event.templateType = kUARetailEventTemplate;
    event.properties = propertyDictionary;
    
    return event;
}

@end

File: Airship/AirshipCore/Source/UARetailEventTemplate.swift
/* Copyright Airship and Contributors */

import Foundation

/**
 * A UARetailEventTemplate represents a custom retail event template for the
 * application.
 */
@objc
public class UARetailEventTemplate: NSObject {
    /**
     * The event's value. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     */
    @objc
    public var eventValue: NSNumber?

    /**
     * The event's transaction ID. The transaction ID's length must not exceed 255
     * characters or it will invalidate the event.
     */
    @objc
    public var transactionID: String?

    /**
     * The event's ID.
     */
    @objc
    public var identifier: String?

    /**
     * The event's category.
     */
    @objc
    public var category: String?

    /**
     * The event's description.
     */
    @objc
    public var eventDescription: String?

    /**
     * The brand..
     */
    @objc
    public var brand: String?

    /**
     * If the item is new or not.
     */
    @objc
    public var isNewItem : Bool {
        get {
            return self._isNewItem ?? false
        }
        set {
            self._isNewItem = newValue
        }
    }

    private var _isNewItem: Bool?
    private let eventName: String
    private let source: String?
    private let medium: String?
    private let wishlistName: String?
    private let wishlistID: String?

    /**
     * Factory method for creating a browsed event template.
     * @returns A Retail event template instance
     */
    @objc(browsedTemplate)
    public class func browsedTemplate() -> UARetailEventTemplate {
        return browsedTemplate(value: nil)
    }

    /**
     * Factory method for creating a browsed event template with a value.
     *
     * @param valueString The value of the event as as string. The value must be between
     * -2^31 and 2^31 - 1 or it will invalidate the event.
     * @returns A Retail event template instance
     */
    @objc(browsedTemplateWithValueFromString:)
    public class func browsedTemplate(valueString: String?) -> UARetailEventTemplate {
        let decimalValue = valueString != nil ? NSDecimalNumber(string: valueString) : nil
        return browsedTemplate(value: decimalValue)
    }

    /**
     * Factory method for creating a browsed event template with a value.
     *
     * @param value The value of the event. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     * @returns A Retail event template instance
     */
    @objc(browsedTemplateWithValue:)
    public class func browsedTemplate(value: NSNumber?) -> UARetailEventTemplate {
        return UARetailEventTemplate("browsed", value: value)
    }

    /**
     * Factory method for creating an addedToCart event template.
     * @returns A Retail event template instance
     */
    @objc(addedToCartTemplate)
    public class func addedToCartTemplate() -> UARetailEventTemplate {
        return addedToCartTemplate(value: nil)
    }

    /**
     * Factory method for creating an addedToCart event template with a value.
     *
     * @param valueString The value of the event as as string. The value must be between
     * -2^31 and 2^31 - 1 or it will invalidate the event.
     * @returns A Retail event template instance
     */
    @objc(addedToCartTemplateWithValueFromString:)
    public class func addedToCartTemplate(valueString: String?) -> UARetailEventTemplate {
        let decimalValue = valueString != nil ? NSDecimalNumber(string: valueString) : nil
        return addedToCartTemplate(value: decimalValue)
    }

    /**
     * Factory method for creating an addedToCart event template with a value.
     *
     * @param value The value of the event. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     * @returns A Retail event template instance
     */
    @objc(addedToCartTemplateWithValue:)
    public class func addedToCartTemplate(value: NSNumber?) -> UARetailEventTemplate {
        return UARetailEventTemplate("added_to_cart", value: value)
    }

    /**
     * Factory method for creating a starredProduct event template
     * @returns A Retail event template instance
     */
    @objc(starredProductTemplate)
    public class func starredProductTemplate() -> UARetailEventTemplate {
        return starredProductTemplate(value: nil)
    }

    /**
     * Factory method for creating a starredProduct event template with a value.
     *
     * @param valueString The value of the event as as string. The value must be between
     * -2^31 and 2^31 - 1 or it will invalidate the event.
     * @returns A Retail event template instance
     */
    @objc(starredProductTemplateWithValueFromString:)
    public class func starredProductTemplate(valueString: String?) -> UARetailEventTemplate {
        let decimalValue = valueString != nil ? NSDecimalNumber(string: valueString) : nil
        return starredProductTemplate(value: decimalValue)
    }

    /**
     * Factory method for creating a starredProduct event template with a value.
     *
     * @param value The value of the event. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     * @returns A Retail event template instance
     */
    @objc(starredProductTemplateWithValue:)
    public class func starredProductTemplate(value: NSNumber?) -> UARetailEventTemplate {
        return UARetailEventTemplate("starred_product", value: value)
    }

    /**
     * Factory method for creating a purchased event template.
     * @returns A Retail event template instance
     */
    @objc(purchasedTemplate)
    public class func purchasedTemplate() -> UARetailEventTemplate {
        return purchasedTemplate(value: nil)
    }

    /**
     * Factory method for creating a purchased event template with a value.
     *
     * @param valueString The value of the event as as string. The value must be between
     * -2^31 and 2^31 - 1 or it will invalidate the event.
     * @returns A Retail event template instance
     */
    @objc(purchasedTemplateWithValueFromString:)
    public class func purchasedTemplate(valueString: String?) -> UARetailEventTemplate {
        let decimalValue = valueString != nil ? NSDecimalNumber(string: valueString) : nil
        return purchasedTemplate(value: decimalValue)
    }

    /**
     * Factory method for creating a purchased event template with a value.
     *
     * @param value The value of the event. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     * @returns A Retail event template instance
     */
    @objc(purchasedTemplateWithValue:)
    public class func purchasedTemplate(value: NSNumber?) -> UARetailEventTemplate {
        return UARetailEventTemplate("purchased", value: value)
    }

    /**
     * Factory method for creating a sharedProduct template event.
     * @returns A Retail event template instance
     */
    @objc(sharedProductTemplate)
    public class func sharedProductTemplate() -> UARetailEventTemplate {
        return sharedProductTemplate(value: nil, source: nil, medium: nil)
    }

    /**
     * Factory method for creating a sharedProduct event template with a value.
     *
     * @param valueString The value of the event as as string. The value must be between
     * -2^31 and 2^31 - 1 or it will invalidate the event.
     * @returns A Retail event template instance
     */
    @objc(sharedProductTemplateWithValueFromString:)
    public class func sharedProductTemplate(valueString: String?) -> UARetailEventTemplate {
        let decimalValue = valueString != nil ? NSDecimalNumber(string: valueString) : nil
        return sharedProductTemplate(value: decimalValue, source: nil, medium: nil)
    }

    /**
     * Factory method for creating a sharedProduct event template with a value.
     *
     * @param value The value of the event. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     * @returns A Retail event template instance
     */
    @objc(sharedProductTemplateWithValue:)
    public class func sharedProductTemplate(value: NSNumber?) -> UARetailEventTemplate {
        return sharedProductTemplate(value: value, source: nil, medium: nil)
    }

    /**
     * Factory method for creating a sharedProduct event template.
     * @param source The source as an NSString.
     * @param medium The medium as an NSString
     * @returns A Retail event template instance.
     */
    @objc(sharedProductTemplateWithSource:withMedium:)
    public class func sharedProductTemplate(source: String?, medium: String?) -> UARetailEventTemplate {
        return sharedProductTemplate(value: nil, source: source, medium: medium)
    }

    /**
     * Factory method for creating a sharedProduct event template with a value.
     *
     * @param valueString The value of the event as as string. The value must be between
     * -2^31 and 2^31 - 1 or it will invalidate the event.
     * @param source The source as an NSString.
     * @param medium The medium as an NSString.
     * @returns A Retail event template instance
     */
    @objc(sharedProductTemplateWithValueFromString:withSource:withMedium:)
    public class func sharedProductTemplate(valueString: String?, source: String?, medium: String?) -> UARetailEventTemplate {
        let decimalValue = valueString != nil ? NSDecimalNumber(string: valueString) : nil
        return sharedProductTemplate(value: decimalValue, source: source, medium: medium)
    }

    /**
     * Factory method for creating a sharedProduct event template with a value.
     *
     * @param value The value of the event. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     * @param source The source as an NSString.
     * @param medium The medium as an NSString.
     * @returns A Retail event template instance
     */
    @objc(sharedProductTemplateWithValue:withSource:withMedium:)
    public class func sharedProductTemplate(value: NSNumber?, source: String?, medium: String?) -> UARetailEventTemplate {
        return UARetailEventTemplate("shared_product", value: value, source: source, medium: medium)
    }

    /**
     * Factory method for creating a wishlist event template.
     * @returns A Retail event template instance
     */
    @objc(wishlistTemplate)
    public class func wishlistTemplate() -> UARetailEventTemplate {
        return wishlistTemplate(name: nil, wishlistID: nil)
    }


    /**
     * Factory method for creating a wishlist event template with a wishlist name and ID.
     *
     * @param name The name of the wishlist as as string.
     * @param wishlistID The ID of the wishlist as as string.
     * @returns A Retail event template instance
     */
    @objc(wishlistTemplateWithName:wishlistID:)
    public class func wishlistTemplate(name: String?, wishlistID: String?) -> UARetailEventTemplate {
        return UARetailEventTemplate("wishlist", wishlistName: name, wishlistID: wishlistID)
    }

    private init(_ eventName: String, value: NSNumber? = nil, source: String? = nil, medium: String? = nil, wishlistName: String? = nil, wishlistID: String? = nil) {
        self.eventName = eventName
        self.eventValue = value
        self.source = source
        self.medium = medium
        self.wishlistID = wishlistID
        self.wishlistName = wishlistName
        super.init()
    }

    /**
     * Creates the custom media event.
     */
    @objc
    public func createEvent() -> UACustomEvent {
        var propertyDictionary: [String : Any] = [:]
        propertyDictionary["ltv"] = self.eventName == "purchased" && self.eventValue != nil
        propertyDictionary["id"] = self.identifier
        propertyDictionary["category"] = self.category
        propertyDictionary["brand"] = self.brand
        propertyDictionary["new_item"] = self._isNewItem
        propertyDictionary["source"] = self.source
        propertyDictionary["medium"] = self.medium
        propertyDictionary["wishlist_name"] = self.wishlistName
        propertyDictionary["wishlist_id"] = self.wishlistID
        propertyDictionary["description"] = self.eventDescription

        let event = UACustomEvent(name: self.eventName)
        event.templateType = "retail"
        event.eventValue = self.eventValue
        event.transactionID = self.transactionID
        event.properties = propertyDictionary
        return event
    }
}
